solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java

```diff
@@ -93,69 +87,71 @@ public class CoreContainerProvider implements ServletContextListener {
   private HttpClient httpClient;
   private SolrMetricManager metricManager;
   private RateLimitManager rateLimitManager;
-  private final CountDownLatch init = new CountDownLatch(1);
   private String registryName;
-  // AFAIK the only reason we need this is to support JettySolrRunner for tests. In tests we might
-  // have multiple CoreContainers in the same JVM, but I *think* that doesn't happen in a real
-  // server.
-  private static final Map<ContextInitializationKey, ServiceHolder> services =
-      Collections.synchronizedMap(new WeakHashMap<>());
-
-  // todo: dependency injection instead, but for now this method and the associated map will have
-  // to suffice.
-  // Note that this relies on ServletContext.equals() not implementing anything significantly
-  // different than Object.equals for its .equals method (I've found no implementation that even
-  // implements it).
-  public static ServiceHolder serviceForContext(ServletContext ctx) throws InterruptedException {
-    ContextInitializationKey key = new ContextInitializationKey(ctx);
-    return services.computeIfAbsent(key, ServiceHolder::new);
+
+  /** Acquires an instance from the context, waiting if necessary. */
+  public static CoreContainerProvider serviceForContext(ServletContext ctx)
```

Comment:
We wait for it; previously did not.  This allows us to return a bona-fide CoreContainerProvider instead of a ServiceHolder wrapper.  Alas, even CCP is itself a wrapper to what we really want.